src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+#include <cstring>
+#include <dlfcn.h>
+
+#include "hooks.h"
+#include "linker_shenanigans.h"
+#include "logging.h"
+#include "macho_utils.h"
+
+#ifndef __APPLE__
+#    error "This file can only be compiled in MacOS systems"
+#endif
+
+namespace memray::linker {
+
+template<typename Hook>
+static void
+patch_symbol(
+        const Hook& hook,
+        typename Hook::signature_t intercept,
+        const char* symname,
+        void* addr,
+        vm_size_t section_size,
+        bool restore_original)
+{
+    vm_size_t trunc_address = trunc_page((vm_size_t)addr);
+    vm_size_t trunc_size = (vm_size_t)addr - trunc_address;
+
+    kern_return_t err = vm_protect(
+            mach_task_self(),
+            (uintptr_t)trunc_address,
+            section_size + trunc_size,
+            0,
+            VM_PROT_READ | VM_PROT_WRITE | VM_PROT_COPY);
+    if (err == KERN_SUCCESS) {
+        auto typedAddr = reinterpret_cast<typename Hook::signature_t*>(addr);
+        *typedAddr = restore_original ? hook.d_original : intercept;
+        LOG_IF_DEBUG(symname << " intercepted!");
+    } else {
+        LOG(ERROR) << "Failed to patch " << symname;
```

Comment:
Should we do anything else here besides log an error?

Reply:
No, we do the same for linux IIRC. There is nothing else we can do if the page cannot be unprotected.